src/index/hgraph_index.cpp

```diff
@@ -0,0 +1,676 @@
+                } else {
+                    bottom_graph_->InsertNeighborsById(inner_id, Vector<InnerIdType>(allocator_));
+                }
+                bottom_graph_->IncreaseTotalCount(1);
+            }
+
+            if (need_lock) {
+                enter_point_id_ = inner_id;
+            }
+        }
+    };
+
+    if (this->build_pool_ != nullptr) {
+        auto task_size = (total + this->build_thread_count_ - 1) / this->build_thread_count_;
+        for (uint64_t j = 0; j < this->build_thread_count_; ++j) {
+            auto end = std::min(j * task_size + task_size, total);
+            this->build_pool_->enqueue(build_func, j * task_size, end);
+        }
+        this->build_pool_->wait_until_nothing_in_flight();
+    } else {
+        build_func(0, total);
+    }
+}
+
+std::shared_ptr<GraphInterface>
+HGraphIndex::generate_one_route_graph() {
+    return std::make_shared<SparseGraphDataCell>(this->allocator_,
+                                                 bottom_graph_->MaximumDegree() / 2);
+}
+
+HGraphIndex::MaxHeap
+HGraphIndex::search_one_graph(const float* query,
+                              const std::shared_ptr<GraphInterface>& graph,
+                              const std::shared_ptr<FlattenInterface>& flatten,
+                              InnerIdType ep,
+                              uint64_t ef,
+                              BaseFilterFunctor* isIdAllowed) const {
+    auto visited_list = this->pool_->getFreeVisitedList();
+
+    auto* visited_array = visited_list->mass;
+    auto visited_array_tag = visited_list->curV;
+    auto computer = flatten->FactoryComputer(query);
+    auto prefetch_neighbor_visit_num = 1;  // TODO(LHT) Optimize the param;
+
+    MaxHeap candidate_set(allocator_);
+    MaxHeap cur_result(allocator_);
+    float dist = 0.0f;
+    auto lower_bound = std::numeric_limits<float>::max();
+    flatten->Query(&dist, computer, &ep, 1);
+    if (not isIdAllowed || (*isIdAllowed)(ep)) {
+        cur_result.emplace(dist, ep);
+        lower_bound = cur_result.top().first;
+    }
+    candidate_set.emplace(-dist, ep);
+    visited_array[ep] = visited_array_tag;
+
+    Vector<InnerIdType> neighbors(allocator_);
+    Vector<InnerIdType> to_be_visited(graph->MaximumDegree(), allocator_);
+    Vector<float> tmp_result(graph->MaximumDegree(), allocator_);
+
+    while (not candidate_set.empty()) {
+        auto current_node_pair = candidate_set.top();
+
+        if ((-current_node_pair.first) > lower_bound && cur_result.size() == ef) {
+            break;
+        }
+        candidate_set.pop();
+
+        auto current_node_id = current_node_pair.second;
+        {
+            std::shared_lock<std::shared_mutex> lock(neighbors_mutex_[current_node_id]);
+            graph->GetNeighbors(current_node_id, neighbors);
+        }
+        if (!neighbors.empty()) {
+            flatten->Prefetch(neighbors[0]);
+#ifdef USE_SSE
```

Comment:
what's the difference between the `flatten->Prefetch` and the following prefetch process.

Reply:
this prefetch vec data, the following prefetch visit_list